### PR TITLE
feat: support console openssl for signing (useful for gost)

### PR DESCRIPTION
**Pull request by kalloc**
hi, let's discuss about the interface

### CODE REVIEW

File: src/Esia/OpenId.php [hunk 1]
```diff
@@ -50,7 +50,8 @@ public function __construct(Config $config, ClientInterface $client = null)
         $this->config = $config;
         $this->client = $client ?? new GuzzleHttpClient(new Client());
         $this->logger = new NullLogger();
-        $this->signer = new SignerPKCS7(
```
Comment (fr05t1k) [inline src/Esia/OpenId.php]: Let's add Esia\Signer\SignerInterface as an optional dependency in a setter

File: src/Esia/Signer/CliSignerPKCS7.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace Esia\Signer;
+
+use Esia\Signer\Exceptions\SignFailException;
+use Psr\Log\LoggerAwareTrait;
+
+
+class CliSignerPKCS7 extends SignerPKCS7
```
Comment (fr05t1k) [inline src/Esia/Signer/CliSignerPKCS7.php]: It's better to keep Signers independent and don't share logic between them.

File: src/Esia/Signer/SignerPKCS7.php [hunk 3]
```diff
@@ -20,28 +20,28 @@ class SignerPKCS7 implements SignerInterface
      *
      * @var string
      */
-    private $certPath;
+    protected $certPath;
```
Comment (fr05t1k) [inline src/Esia/Signer/SignerPKCS7.php]: Keep properties as private

File: tests/unit/OpenIdTest.php [hunk 4]
```diff
@@ -170,3 +170,28 @@ private function buildClientWithResponses(array $responses): ClientInterface
         return new GuzzleHttpClient($guzzleClient);
     }
 }
+
+
+
+class OpenIdCliOpensslTest extends OpenIdTest {
```
Comment (fr05t1k) [inline tests/unit/OpenIdTest.php]: it's better to keep classes in separate files

File: src/Esia/Config.php [hunk 5]
```diff
@@ -66,6 +70,7 @@ public function __construct(array $config = [])
         if (!$this->certPath) {
             throw new InvalidConfigurationException('Please provide certPath');
         }
+        $this->signerClass = $config['signerClass'] ?? '\Esia\Signer\SignerPKCS7';
```
Comment (fr05t1k) [inline src/Esia/Config.php]: Don't need to specify class in the config if we use it as a dependency for `OpenId` class

Comment (kalloc) [reply src/Esia/Config.php]: @fr05t1k May I ask you to show what do you mean in code?

File: src/Esia/Signer/SignerPKCS7.php (hunk 3 above)
Comment (kalloc) [reply src/Esia/Signer/SignerPKCS7.php]: Did you mean that CliSignerPKCS7 should have private `$certPath` and another?
It looks like copy & past

File: src/Esia/Signer/CliSignerPKCS7.php (hunk 2 above)
Comment (kalloc) [reply src/Esia/Signer/CliSignerPKCS7.php]: but what about certPath, checkFilesExists, getRandomString, and urlSafe?

Comment (fr05t1k) [reply src/Esia/Signer/CliSignerPKCS7.php]: Let's create  AbstractSignerPKCS7 where we can store common logic.

File: src/Esia/Config.php (hunk 5 above)
Comment (fr05t1k) [reply src/Esia/Config.php]:
```
$config = new Config();
$openId = new OpenId($config);
$openId->setSigner(new CliSignerPKCS7());
```
In this case we can avoid dynamic class creation and allow to config Signer separately.

**Comment by kalloc** [comment]
I have no idea how-to combine AbstractSignerPKCS7 and private *path, any ideas?

**Comment by kalloc** [comment]
@fr05t1k thanks

**Comment by fr05t1k** [comment]
@kalloc thanks ;)

**Pull request closed without merging**